File: server/src/user/user.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';
import { Provider } from 'src/generated/prisma/enums';
import { UpdateProfileDto } from './dto/update-profile.dto';

@Injectable()
export class UserService {
  constructor(private prisma: PrismaService) {}

  findByEmail(email: string) {
    return this.prisma.user.findUnique({ where: { email } });
  }

  findById(id: string) {
    return this.prisma.user.findUnique({ where: { id } });
  }

  create(data: { email: string; name: string; hashedPassword: string; provider?: Provider }) {
    return this.prisma.user.create({ data });
  }

  async findOrCreateGoogleUser(data: { email: string; name: string; profileSrc: string }) {
    const existing = await this.findByEmail(data.email);
    if (existing) return existing;

    return this.prisma.user.create({
      data: {
        email: data.email,
        name: data.name,
        profileSrc: data.profileSrc,
        provider: 'GOOGLE',
      },
    });
  }

  async getProfile(userId: string) {
    const user = await this.prisma.user.findUnique({
      where: { id: userId },
      select: {
        id: true,
        name: true,
        email: true,
        bio: true,
        websiteUrl: true,
        profileSrc: true,
        createdAt: true,
        _count: {
          select: {
            posts: true,
            followers: true,
            following: true,
          },
        },
      },
    });

    if (!user) throw new NotFoundException('User not found');
    return user;
  }

  async updateProfile(userId: string, dto: UpdateProfileDto) {
    return this.prisma.user.update({
      where: { id: userId },
      data: dto,
      select: {
        id: true,
        name: true,
        email: true,
        bio: true,
        websiteUrl: true,
        profileSrc: true,
      },
    });
  }

  async getSavedPosts(userId: string) {
    return this.prisma.save.findMany({
      where: { userId },
      orderBy: { createdAt: 'desc' },
      include: {
        post: {
          include: {
            author: { select: { id: true, name: true, profileSrc: true } },
            _count: { select: { likes: true, comments: true, ratings: true } },
          },
        },
      },
    });
  }

  async saveInterests(userId: string, interests: string[]) {
    await this.prisma.userInterest.deleteMany({ where: { userId } });
    return this.prisma.userInterest.createMany({
      data: interests.map((interest) => ({ userId, interest })),
    });
  }

  async getInterests(userId: string) {
    return this.prisma.userInterest.findMany({
      where: { userId },
      select: { interest: true },
    });
  }
}

File: client/lib/services/user_service.dart
import 'package:client/data/models/api_user.dart';
import 'package:client/services/api_client.dart';

class UserService {
  String? currentUserId;
  ApiUser? cachedUser;

  Future<ApiUser> getMyProfile() async {
    final response = await dio.get('/api/users/me');
    final user = ApiUser.fromJson(response.data);
    currentUserId = user.id;
    cachedUser = user;
    return user;
  }

  Future<List<String>> getInterests() async {
    final response = await dio.get('/api/users/me/interests');
    final List data = response.data;
    return data.map((item) => item['interest'] as String).toList();
  }

  Future<void> saveInterests(List<String> interests) async {
    await dio.post('/api/users/me/interests', data: {'interests': interests});
  }

  Future<void> updateProfile({
    required String name,
    String? bio,
    String? websiteUrl,
    String? profileSrc,
  }) async {
    await dio.patch('/api/users/me', data: {
      'name': name,
      if (bio != null) 'bio': bio,
      if (websiteUrl != null) 'websiteUrl': websiteUrl,
      if (profileSrc != null) 'profileSrc': profileSrc,
    });
  }

  Future<ApiUser> getUserProfile(String userId) async {
    final response = await dio.get('/api/users/$userId');
    return ApiUser.fromJson(response.data);
  }
}

final userService = UserService();